transformer_engine/common/permutation/permutation.cu

```diff
@@ -0,0 +1,296 @@
+      }
+    } else {
+      // Unpermute bwd with probs
+      int blocks = num_rows;
+      int threads = 32;
+      size_t smem_bytes = num_topK * sizeof(TCompute);
+
+      if (num_topK <= 8) {
+        moe_permute_kernel<T, TCompute, 8, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 16) {
+        moe_permute_kernel<T, TCompute, 16, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 32) {
+        moe_permute_kernel<T, TCompute, 32, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 64) {
+        moe_permute_kernel<T, TCompute, 64, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else if (num_topK <= 128) {
+        moe_permute_kernel<T, TCompute, 128, true><<<blocks, threads, smem_bytes, stream>>>(
+            input, input_fwd, output, prob, prob_grad, row_id_map, num_rows, num_topK, num_cols);
+      } else {
+        throw std::runtime_error("num_topK cannot exceed 128.");
+      }
+    }
+  } else {
+    int blocks = num_rows;
+    int threads = std::min(num_cols / kElementsPerAccess, 1024);
+    size_t smem_bytes = num_topK * sizeof(TCompute);
+
+    if (prob == nullptr) {
+      // Permute bwd
+      // Unpermute fwd without probs
+      moe_unpermute_kernel<T, TCompute, false><<<blocks, threads, smem_bytes, stream>>>(
+          input, output, row_id_map, prob, num_rows, num_topK, num_cols);
+    } else {
+      // Unpermute fwd with probs
+      moe_unpermute_kernel<T, TCompute, true><<<blocks, threads, smem_bytes, stream>>>(
+          input, output, row_id_map, prob, num_rows, num_topK, num_cols);
+    }
+  }
+}
+
+#define FUNCTION_INSTANTIATION(T, FWD)                                               \
+  template void moe_permutation_launcher<T, FWD>(                                    \
+      const void *input, void *output, const int *sorted_row_id, int *row_id_map,    \
+      const float *prob, const int num_rows, const int num_topK, const int num_cols, \
+      const int num_out_tokens, cudaStream_t stream, float *prob_grad, const void *input_fwd);
+
+FUNCTION_INSTANTIATION(float, true)
+FUNCTION_INSTANTIATION(float, false)
+FUNCTION_INSTANTIATION(half, true)
+FUNCTION_INSTANTIATION(half, false)
+FUNCTION_INSTANTIATION(__nv_bfloat16, true)
+FUNCTION_INSTANTIATION(__nv_bfloat16, false)
+FUNCTION_INSTANTIATION(__nv_fp8_e5m2, true)
+FUNCTION_INSTANTIATION(__nv_fp8_e5m2, false)
+FUNCTION_INSTANTIATION(__nv_fp8_e4m3, true)
+FUNCTION_INSTANTIATION(__nv_fp8_e4m3, false)
```

Comment:
```suggestion

```

Reply:
Hi, please correct me if I'm wrong. I don't think we need this instantiation.

Reply:
Maybe yes, because we declared the fucntion with template in permutation.h file. The compiler dont know the template paremeters needed when instantiate the function, so that it chose not to instantiate any functions. This will lead to a link error at runtime.